### PR TITLE
Add/improve likely/unlikely hinting to all deflate algorithms.

### PR DESCRIPTION
**Pull request by Dead2**
Overall performance improvement shows ~0.42% speed increase.

This was based on coverage data, manual interpretation and for deflate_medium; a good deal of experimentation.

### CODE REVIEW

**Comment by Dead2** [comment]
## Deflatebench config
```
Activated multiple generated file mode. Source: silesia-small.tar
Level 1: /tmp/deflatebench-1.tmp 270.1 MiB / 283,253,760 B
Level 2: /tmp/deflatebench-2.tmp 135.1 MiB / 141,626,880 B
Level 3: /tmp/deflatebench-3.tmp 105.1 MiB / 110,154,240 B
Level 4: /tmp/deflatebench-4.tmp 90.0 MiB / 94,417,920 B
Level 5: /tmp/deflatebench-5.tmp 90.0 MiB / 94,417,920 B
Level 6: /tmp/deflatebench-6.tmp 75.0 MiB / 78,681,600 B
Level 7: /tmp/deflatebench-7.tmp 60.0 MiB / 62,945,280 B
Level 8: /tmp/deflatebench-8.tmp 45.0 MiB / 47,208,960 B
Level 9: /tmp/deflatebench-9.tmp 45.0 MiB / 47,208,960 B

 Runs: 70
 Levels: 1-9
 Trimworst: 30
```

## Baseline 9609cb56a8f62868ccf264493bc9c3b4d5762fcf
```
   text    data     bss     dec     hex filename
  95640    1472      32   97144   17b78 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.870%    1.881/1.909/1.922       1.159/1.223/1.239  163918126
 2      43.903%    1.826/1.856/1.870       0.616/0.637/0.649  62178636
 3      42.510%    1.702/1.727/1.736       0.463/0.478/0.487  46826710
 4      41.490%    1.690/1.707/1.716       0.388/0.404/0.412  39174432
 5      41.215%    1.840/1.858/1.868       0.366/0.400/0.409  38913920
 6      41.038%    1.839/1.853/1.862       0.305/0.329/0.338  32289128
 7      40.784%    2.003/2.020/2.027       0.253/0.268/0.273  25671760
 8      40.707%    1.942/1.954/1.960       0.186/0.196/0.202  19217358
 9      40.696%    2.158/2.170/2.177       0.175/0.198/0.206  19212387

 avg1   43.357%                1.895                   0.459
 tot                         682.131                 165.354
```

## PR/670 c1572bd1f35eba6d8f97c04f4b957f1fb85da1fa
```
   text    data     bss     dec     hex filename
  95320    1472      32   96824   17a38 libz-ng.so.1

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.870%    1.885/1.908/1.922       1.181/1.213/1.228  163918126
 2      43.903%    1.820/1.851/1.865       0.610/0.631/0.644  62178636
 3      42.510%    1.701/1.725/1.732       0.456/0.480/0.492  46826710
 4      41.490%    1.673/1.693/1.703       0.374/0.399/0.407  39174432
 5      41.215%    1.825/1.846/1.852       0.382/0.401/0.409  38913920
 6      41.038%    1.823/1.840/1.848       0.311/0.330/0.337  32289128
 7      40.784%    1.995/2.006/2.014       0.249/0.266/0.272  25671760
 8      40.707%    1.927/1.943/1.950       0.191/0.198/0.205  19217358
 9      40.696%    2.142/2.158/2.165       0.184/0.197/0.204  19212387

 avg1   43.357%                1.886                   0.457
 tot                         678.841                 164.558

 Level   Comp   Comptime min/avg/max  Decomptime min/avg/max  Compressed size
 1      57.870%    1.890/1.911/1.927       1.187/1.211/1.227  163918126
 2      43.903%    1.815/1.854/1.869       0.611/0.632/0.644  62178636
 3      42.510%    1.706/1.723/1.733       0.459/0.479/0.489  46826710
 4      41.490%    1.674/1.694/1.702       0.387/0.401/0.408  39174432
 5      41.215%    1.831/1.845/1.854       0.385/0.399/0.409  38913920
 6      41.038%    1.824/1.841/1.850       0.313/0.330/0.337  32289128
 7      40.784%    1.984/2.007/2.015       0.252/0.264/0.272  25671760
 8      40.707%    1.935/1.944/1.951       0.185/0.198/0.204  19217358
 9      40.696%    2.146/2.161/2.166       0.184/0.196/0.201  19212387

 avg1   43.357%                1.887                   0.457
 tot                         679.257                 164.360
```

File: deflate_medium.c [hunk 1]
```diff
@@ -273,15 +273,15 @@ ZLIB_INTERNAL block_state deflate_medium(deflate_state *s, int flush) {
         /* move the "cursor" forward */
         s->strstart += current_match.match_length;
 
-        if (bflush)
+        if (UNLIKELY(bflush))
```
Comment (nmoinvaz) [inline deflate_medium.c:276]: Is it ever possible that when `bflush=1` this will never get hit?

File: deflate_medium.c [hunk 2]
```diff
@@ -182,7 +182,7 @@ ZLIB_INTERNAL block_state deflate_medium(deflate_state *s, int flush) {
             if (s->lookahead < MIN_LOOKAHEAD && flush == Z_NO_FLUSH) {
```
Comment (nmoinvaz) [inline deflate_medium.c:182]: Why not `UNLIKELY` here too?

File: deflate_medium.c [hunk 3]
```diff
@@ -273,15 +273,15 @@ ZLIB_INTERNAL block_state deflate_medium(deflate_state *s, int flush) {
         /* move the "cursor" forward */
         s->strstart += current_match.match_length;
 
-        if (bflush)
+        if (UNLIKELY(bflush))
             FLUSH_BLOCK(s, 0);
     }
     s->insert = s->strstart < MIN_MATCH-1 ? s->strstart : MIN_MATCH-1;
     if (flush == Z_FINISH) {
```
Comment (nmoinvaz) [inline deflate_medium.c:280]: Why not `UNLIKELY` here too?

**Review by nmoinvaz — COMMENTED** [review]
Provided some feedback..

File: deflate_medium.c (hunk 1 above)
Comment (Dead2) [reply deflate_medium.c:276]: The question is rather how often do we call this with bflush=1? And do we want to optimize for that?

File: deflate_medium.c (hunk 3 above)
Comment (Dead2) [reply deflate_medium.c:280]: It hurts short compressions, and also hurts performance in my huge-file tests. As I mentioned, deflate_medium needed a lot of testing.

File: deflate_medium.c (hunk 2 above)
Comment (Dead2) [reply deflate_medium.c:182]: I am not sure, but benchmark-wise it hurt performance.

**Comment by Dead2** [comment]
The reason some of those don't like being `UNLIKELY` branches might be that they get exported by the compiler as separate blocks, and then the linker de-duplicates those blocks, so the unlikely branch then needs to access memory far away out of cache.

At least I suspect that is why the zlib-ng library shrinks by 320bytes with this PR. Originally it was ~500 bytes.

File: deflate_medium.c (hunk 1 above)
Comment (nmoinvaz) [reply deflate_medium.c:276]: I see ok

File: deflate_medium.c (hunk 3 above)
Comment (nmoinvaz) [reply deflate_medium.c:280]: Hmm I wonder also about doing: s->insert = UNLIKELY(s->strstart < MIN_MATCH-1) ? s->strstart : MIN_MATCH-1;

File: deflate_medium.c (hunk 2 above)
Comment (nmoinvaz) [reply deflate_medium.c:182]: Ok sounds good

**Comment by nmoinvaz** [comment]
Wow that is pretty cool. I haven’t had much experience with those builtins in the past.

**Comment by Dead2** [comment]
If you are interested, I can recommend experimenting with Intel System Studio, at least if you have an Intel CPU.
Intel- `Advisor`, `Vtune Profiler` and `Inspector` are tools that let you dive deep into what the code does down at the cpu-level.
It lets you know how well you utilize parallelism in vectorized code and how badly you are cache-bound for example.

Analyzing the output from those tools is demanding though, and to really get good use of it you might need to read up on how the cpu pipelines work and how it interacts with the uncore caches etc.

If you are up for it, this webpage and especially the video at the bottom gives a really good into on the subject. (Youtube claims I watched this in 2016, has it really been that long?)
https://medium.com/software-design/why-software-developers-should-care-about-cpu-caches-8da04355bb8a

Another (long) talk on the subject:
https://www.youtube.com/watch?v=L7zSU9HI-6I

And if you want to watch something completely unrelated but truly awesome and enjoyable IMO:
https://www.youtube.com/watch?v=zBkNBP00wJE

**Comment by gdevenyi** [comment]
There's also the option of using GCC's gprof to do similar analysis of codepaths: https://www.thegeekstuff.com/2012/08/gprof-tutorial/

**Pull request merged**